clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -1144,7 +1149,11 @@ public void close() {
         public SelectorMetrics(Metrics metrics, String metricGrpPrefix, Map<String, String> metricTags, boolean metricsPerConnection) {
             this.metrics = metrics;
             this.metricTags = metricTags;
-            this.metricsPerConnection = metricsPerConnection;
+            if (metricsPerConnection) {
+                this.connectionMetrics = new ConcurrentHashMap<>();
```

Comment:
Do we really need a `ConcurrentHashMap` here? My understanding is that the selector should be used by a single thread. Isn't it the case here?

Reply:
Agreed. The Selector is advertised not to be thread safe.

Reply:
Oh yes, its javadoc states that `This class is not thread safe!`;